### ISSUE TITLE
Monitor looses focus after `WorkspaceManagerActivateAdjacent`

### ISSUE THREAD

**Issue by urob** [dalyIsaac/Whim#821]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Behavior

Triggering `WorkspaceManager.ActivateAdjacent` via keybinds on an empty monitor sometimes results in the focused monitor loosing focus. This is a regression from some time around the butler-refactor. (Unfortunately, I can't remember the exact commit when this started).

## Example

Initially active workspaces:
- `Workspace 3` on the primary monitor ("`DISPLAY`") -- focused 
- `Workspace 1` on the secondary monitor ("`DISPLAY1`") -- not focused

Final situation after the keybind was pressed:
- `Workspace 4` on the primary monitor ("`DISPLAY`") -- not focused
- `Workspace 1` on the secondary monitor ("`DISPLAY1`") -- focused

I am attaching logs below. For illustrative purposes, the logs are split in two parts:

Part 1 covers everything from when the keybind was first pressed, to the successful switching of the workspace, to just before the previously active monitor looses focus the first time. 

Part 2 show everything after the previously active monitor (`DISPLAY`) first looses focus. There is a bit of ping-pong, but the end result is that the other monitor (`DISPLAY1`) ends up with the focus.

## Logs: Part 1

```
2024-02-05 21:57:40.153 -05:00 [DBG] KeybindManager:82              [GetCommands]                  Getting commands for keybind 'LAlt + H'
2024-02-05 21:57:40.153 -05:00 [DBG] Command:52                     [TryExecute]                   Trying to execute command whim.custom.activate_next_workspace
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerChores:80                [ActivateAdjacent]             Activating next workspace
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerPantry:70                [GetMonitorForWorkspace]       Could not find monitor for workspace 4
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerChores:18                [Activate]                     Activating workspace 4 in monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.153 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.154 -05:00 [DBG] ButlerPantry:70                [GetMonitorForWorkspace]       Could not find monitor for workspace 4
2024-02-05 21:57:40.154 -05:00 [DBG] ButlerPantry:124               [SetMonitorWorkspace]          Setting workspace 4 to monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.154 -05:00 [DBG] Workspace:446                  [Deactivate]                   Deactivating workspace 3
2024-02-05 21:57:40.154 -05:00 [DBG] Window:86                      [Hide]                         awesome(1) — Mozilla Firefox (firefox) [64312] <MozillaWindowClass> {0x002808CC}
2024-02-05 21:57:40.154 -05:00 [DBG] NativeManager:59               [HideWindow]                   Hiding window HWND 0x002808CC
2024-02-05 21:57:40.156 -05:00 [DBG] Workspace:464                  [DoLayout]                     Workspace 4
2024-02-05 21:57:40.156 -05:00 [DBG] DeferWorkspacePosManager:23    [DoLayout]                     Layout 4
2024-02-05 21:57:40.156 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.156 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.156 -05:00 [DBG] DeferWorkspacePosManager:39    [DoLayout]                     Starting layout for workspace 4
2024-02-05 21:57:40.156 -05:00 [DBG] FocusIndicatorPlugin:129       [Hide]                         Hiding focus indicator
2024-02-05 21:57:40.157 -05:00 [DBG] NativeManager:59               [HideWindow]                   Hiding window HWND 0x002509E6
2024-02-05 21:57:40.158 -05:00 [DBG] DeferWorkspacePosManager:50    [SetWindowPos]                 Setting window positions for workspace 4
2024-02-05 21:57:40.158 -05:00 [DBG] SliceLayoutEngine:145          [DoLayout]                     Doing layout on (X: 6, Y: 48, Width: 3828, Height: 2106) on Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True], with 0 windows and 1 minimized windows
2024-02-05 21:57:40.159 -05:00 [DBG] DeferWorkspacePosManager:59    [SetWindowPos]                 Window TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8} has rectangle (X: 0, Y: 0, Width: 0, Height: 0)
2024-02-05 21:57:40.159 -05:00 [DBG] DeferWindowPosHandle:46        [.ctor]                        Creating new WindowDeferPosHandle
2024-02-05 21:57:40.159 -05:00 [DBG] DeferWindowPosHandle:120       [Dispose]                      Disposing WindowDeferPosHandle
2024-02-05 21:57:40.159 -05:00 [DBG] DeferWindowPosHandle:146       [Dispose]                      Setting window position 2 times for 0 windows
2024-02-05 21:57:40.159 -05:00 [DBG] NativeManager:77               [MinimizeWindow]               Minimizing window HWND 0x00E511B8
2024-02-05 21:57:40.174 -05:00 [DBG] NativeManager:77               [MinimizeWindow]               Minimizing window HWND 0x00E511B8
2024-02-05 21:57:40.176 -05:00 [DBG] DeferWindowPosHandle:182       [Dispose]                      Finished setting window position
2024-02-05 21:57:40.176 -05:00 [DBG] FocusIndicatorPlugin:93        [Show]                         Showing focus indicator
2024-02-05 21:57:40.176 -05:00 [DBG] WorkspaceManager:67            [ActiveWorkspace]              Getting active workspace for monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.176 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.176 -05:00 [DBG] FocusIndicatorWindow.xaml:32   [Activate]                     Activating focus indicator window
2024-02-05 21:57:40.176 -05:00 [DBG] DeferWindowPosHandle:46        [.ctor]                        Creating new WindowDeferPosHandle
2024-02-05 21:57:40.176 -05:00 [DBG] DeferWindowPosHandle:89        [DeferWindowPos]               Adding window Whim Focus Indicator (Whim.Runner) [46728] <WinUIDesktopWin32WindowClass> {0x002509E6} after 0x00E511B8 with flags SWP_NOREDRAW, SWP_NOACTIVATE
2024-02-05 21:57:40.176 -05:00 [DBG] DeferWindowPosHandle:120       [Dispose]                      Disposing WindowDeferPosHandle
2024-02-05 21:57:40.176 -05:00 [DBG] DeferWindowPosHandle:146       [Dispose]                      Setting window position 2 times for 1 windows
2024-02-05 21:57:40.176 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.180 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.183 -05:00 [DBG] DeferWindowPosHandle:182       [Dispose]                      Finished setting window position
2024-02-05 21:57:40.183 -05:00 [DBG] Workspace:120                  [FocusLastFocusedWindow]       Focusing last focused window in workspace 4
2024-02-05 21:57:40.183 -05:00 [DBG] Window:68                      [Focus]                        TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.193 -05:00 [DBG] WindowManager:370              [OnWindowFocused]              Window focused: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.193 -05:00 [DBG] MonitorManager:78              [WindowFocused]                Focusing hwnd 0x00E511B8
```

## Logs: Part 2

```
2024-02-05 21:57:40.193 -05:00 [DBG] MonitorManager:89              [WindowFocused]                Setting active monitor to Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.193 -05:00 [DBG] ButlerEventHandlers:127        [WindowManager_WindowFocused]  Window focused: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.193 -05:00 [DBG] Workspace:81                   [WindowFocused]                Focused window TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8} in workspace 4
2024-02-05 21:57:40.193 -05:00 [DBG] ButlerPantry:82                [GetWorkspaceForWindow]        Getting workspace for window TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.193 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.193 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.193 -05:00 [DBG] FocusIndicatorPlugin:93        [Show]                         Showing focus indicator
2024-02-05 21:57:40.194 -05:00 [DBG] WorkspaceManager:67            [ActiveWorkspace]              Getting active workspace for monitor Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.194 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.194 -05:00 [DBG] FocusIndicatorWindow.xaml:32   [Activate]                     Activating focus indicator window
2024-02-05 21:57:40.194 -05:00 [DBG] DeferWindowPosHandle:46        [.ctor]                        Creating new WindowDeferPosHandle
2024-02-05 21:57:40.194 -05:00 [DBG] DeferWindowPosHandle:89        [DeferWindowPos]               Adding window Whim Focus Indicator (Whim.Runner) [46728] <WinUIDesktopWin32WindowClass> {0x002509E6} after 0x0001059E with flags SWP_NOREDRAW, SWP_NOACTIVATE
2024-02-05 21:57:40.194 -05:00 [DBG] DeferWindowPosHandle:120       [Dispose]                      Disposing WindowDeferPosHandle
2024-02-05 21:57:40.194 -05:00 [DBG] DeferWindowPosHandle:146       [Dispose]                      Setting window position 2 times for 1 windows
2024-02-05 21:57:40.195 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.198 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.199 -05:00 [DBG] DeferWindowPosHandle:182       [Dispose]                      Finished setting window position
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:50                [GetMonitorForWindow]          Getting monitor for window: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:50                [GetMonitorForWindow]          Getting monitor for window: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.199 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True]
2024-02-05 21:57:40.202 -05:00 [DBG] WindowManager:279              [WinEventProc]                 Windows event 0x8003 for awesome(1) — Mozilla Firefox (firefox) [64312] <MozillaWindowClass> {0x002808CC}
2024-02-05 21:57:40.202 -05:00 [DBG] WindowManager:384              [OnWindowHidden]               Window hidden: awesome(1) — Mozilla Firefox (firefox) [64312] <MozillaWindowClass> {0x002808CC}
2024-02-05 21:57:40.202 -05:00 [DBG] ButlerPantry:50                [GetMonitorForWindow]          Getting monitor for window: awesome(1) — Mozilla Firefox (firefox) [64312] <MozillaWindowClass> {0x002808CC}
2024-02-05 21:57:40.202 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 3
2024-02-05 21:57:40.202 -05:00 [DBG] ButlerPantry:70                [GetMonitorForWorkspace]       Could not find monitor for workspace 3
2024-02-05 21:57:40.202 -05:00 [DBG] WindowManager:388              [OnWindowHidden]               Window awesome(1) — Mozilla Firefox (firefox) [64312] <MozillaWindowClass> {0x002808CC} is not tracked in a monitor, ignoring event
2024-02-05 21:57:40.202 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.202 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.202 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x002509E6
2024-02-05 21:57:40.203 -05:00 [DBG] WindowManager:279              [WinEventProc]                 Windows event 0x8002 for TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.203 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.203 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.203 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x010F0A2C
2024-02-05 21:57:40.203 -05:00 [DBG] WindowManager:123              [CreateWindow]                 Adding window 0x010F0A2C
2024-02-05 21:57:40.203 -05:00 [DBG] WindowManager:279              [WinEventProc]                 Windows event 0x800B for TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.204 -05:00 [DBG] WindowManager:545              [OnWindowMoved]                Window moved: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.204 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.204 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.204 -05:00 [DBG] WindowManager:279              [WinEventProc]                 Windows event 0x800B for TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.204 -05:00 [DBG] WindowManager:545              [OnWindowMoved]                Window moved: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.204 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.204 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.204 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x002509E6
2024-02-05 21:57:40.205 -05:00 [DBG] WindowManager:279              [WinEventProc]                 Windows event 0x0003 for TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.205 -05:00 [DBG] WindowManager:370              [OnWindowFocused]              Window focused: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.205 -05:00 [DBG] MonitorManager:78              [WindowFocused]                Focusing hwnd 0x00E511B8
2024-02-05 21:57:40.205 -05:00 [DBG] MonitorManager:89              [WindowFocused]                Setting active monitor to Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.205 -05:00 [DBG] ButlerEventHandlers:127        [WindowManager_WindowFocused]  Window focused: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.205 -05:00 [DBG] Workspace:81                   [WindowFocused]                Focused window TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8} in workspace 4
2024-02-05 21:57:40.205 -05:00 [DBG] ButlerPantry:82                [GetWorkspaceForWindow]        Getting workspace for window TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.205 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.205 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.205 -05:00 [DBG] FocusIndicatorPlugin:93        [Show]                         Showing focus indicator
2024-02-05 21:57:40.205 -05:00 [DBG] WorkspaceManager:67            [ActiveWorkspace]              Getting active workspace for monitor Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.205 -05:00 [DBG] ButlerPantry:76                [GetWorkspaceForMonitor]       Getting workspace for monitor Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.205 -05:00 [DBG] FocusIndicatorWindow.xaml:32   [Activate]                     Activating focus indicator window
2024-02-05 21:57:40.205 -05:00 [DBG] DeferWindowPosHandle:46        [.ctor]                        Creating new WindowDeferPosHandle
2024-02-05 21:57:40.205 -05:00 [DBG] DeferWindowPosHandle:89        [DeferWindowPos]               Adding window Whim Focus Indicator (Whim.Runner) [46728] <WinUIDesktopWin32WindowClass> {0x002509E6} after 0x0001059E with flags SWP_NOREDRAW, SWP_NOACTIVATE
2024-02-05 21:57:40.205 -05:00 [DBG] DeferWindowPosHandle:120       [Dispose]                      Disposing WindowDeferPosHandle
2024-02-05 21:57:40.206 -05:00 [DBG] DeferWindowPosHandle:146       [Dispose]                      Setting window position 2 times for 1 windows
2024-02-05 21:57:40.206 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.206 -05:00 [DBG] NativeManager:83               [ShowWindowNoActivate]         Showing window HWND 0x002509E6 no activate
2024-02-05 21:57:40.207 -05:00 [DBG] DeferWindowPosHandle:182       [Dispose]                      Finished setting window position
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:50                [GetMonitorForWindow]          Getting monitor for window: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:50                [GetMonitorForWindow]          Getting monitor for window: TIDAL (TIDAL) [57732] <Chrome_WidgetWin_1> {0x00E511B8}
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:58                [GetMonitorForWorkspace]       Getting monitor for workspace 4
2024-02-05 21:57:40.207 -05:00 [DBG] ButlerPantry:65                [GetMonitorForWorkspace]       Found monitor Monitor[Bounds=(X: 0, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 0, Y: 0, Width: 3840, Height: 2160) Name=DISPLAY ScaleFactor=150 IsPrimary=True] for workspace 4
2024-02-05 21:57:40.207 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.207 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.207 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x00090F80
2024-02-05 21:57:40.208 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x002509E6
2024-02-05 21:57:40.209 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x000100C2
2024-02-05 21:57:40.209 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x000100F4
2024-02-05 21:57:40.210 -05:00 [DBG] WindowManager:133              [CreateWindow]                 Created window Whim Bar (Whim.Runner) [46728] <WinUIDesktopWin32WindowClass> {0x010F0A2C}
2024-02-05 21:57:40.212 -05:00 [DBG] WindowManager:264              [WinEventProc]                 Window 0x010F0A2C with event 0x0003 was ignored, but still notifying listeners of focus
2024-02-05 21:57:40.212 -05:00 [DBG] WindowManager:370              [OnWindowFocused]              Window focused: 
2024-02-05 21:57:40.212 -05:00 [DBG] MonitorManager:78              [WindowFocused]                Focusing hwnd 0x00E511B8
2024-02-05 21:57:40.212 -05:00 [DBG] MonitorManager:89              [WindowFocused]                Setting active monitor to Monitor[Bounds=(X: 3840, Y: 0, Width: 3840, Height: 2160) WorkingArea=(X: 3840, Y: 0, Width: 3840, Height: 2160) Name=\\.\DISPLAY1 ScaleFactor=150 IsPrimary=False]
2024-02-05 21:57:40.212 -05:00 [DBG] ButlerEventHandlers:127        [WindowManager_WindowFocused]  Window focused: 
2024-02-05 21:57:40.212 -05:00 [DBG] FocusIndicatorPlugin:129       [Hide]                         Hiding focus indicator
2024-02-05 21:57:40.212 -05:00 [DBG] NativeManager:59               [HideWindow]                   Hiding window HWND 0x002509E6
2024-02-05 21:57:40.213 -05:00 [DBG] DeferWindowPosManager:62       [RecoverLayout]                Attempting to recover layout
2024-02-05 21:57:40.213 -05:00 [DBG] DeferWindowPosManager:71       [RecoverLayout]                No windows to recover layout for
2024-02-05 21:57:40.213 -05:00 [DBG] WindowManager:321              [AddWindow]                    Adding window 0x002509E6
```

### Config

```csharp
// here's the definition of the command triggered by the keybind:
    context.CommandManager.Add(
        identifier: "activate_previous_workspace",
        title: "Activate the previous inactive workspace",
        callback: () => context.WorkspaceManager.ActivateAdjacent(reverse: true, skipActive: true)
    );
```

### Environment

- Windows:
- Architecture:
- Whim: current main

### Steps To Reproduce

see above

**Comment by dalyIsaac** [comment]
From reading the logs you provided, the Bar received an event from Windows, so the focus was updated. Unfortunately, it's not clear which event was the trigger.

I've added a number of mitigations in #827 which should improve the consistency of the focus when switching workspaces.

**Comment by dalyIsaac** [comment]
I've been able to replicate this too, thanks for the logs

**Comment by dalyIsaac** [comment]
- [x] verified the reproduction steps failed in `main` prior to #829 being merged
- [x] verified that the reproduction steps when running in #829 did not result in monitor focus changing unnecessarily

Hopefully this addresses the issues here 🤞

**Comment by dalyIsaac** [comment]
It's good to hear that it's slowly getting better. For some of the behavior I've been making hypotheses on Windows' actions in different scenarios and accounting for that. The latest example is a good case - `workspace 1` and `Tmux` aren't mentioned until after Whim has received an `EVENT_SYSTEM_FOREGROUND` for the `Tmux` window from Windows. I'll definitely keep an eye on this too.

Thanks for the really good reproduction steps and logs!

**Comment by dalyIsaac** [comment]
That specific issue should be fixed as part of the #835.
